packages/pluggableWidgets/carousel-web/src/Carousel.tsx

```diff
@@ -0,0 +1,39 @@
+import { createElement, ReactNode, useEffect, useState, useCallback, ReactElement } from "react";
+import { executeAction } from "@mendix/piw-utils-internal";
+import { CarouselContainerProps } from "../typings/CarouselProps";
+import { Carousel as CarouselComponent } from "./components/Carousel";
+import loadingCircleSvg from "./ui/loading-circle.svg";
+import "./ui/Carousel.scss";
+
+export function Carousel(props: CarouselContainerProps): ReactNode {
+    const { dataSource, content, showPagination, loop, tabIndex, navigation, animation, delay, autoplay } = props;
+    const [loading, setLoading] = useState(true);
+    const onClick = useCallback(() => executeAction(props.onClickAction), [props.onClickAction]);
+    useEffect(() => {
+        if (dataSource?.status === "available") {
+            setLoading(false);
+        }
+    }, [dataSource]);
+
+    const itemRenderer = useCallback((renderWrapper, item) => renderWrapper(content?.get(item)), [content]);
+    const renderCarousel = (): ReactElement => {
+        return (
+            <CarouselComponent
+                pagination={showPagination}
+                loop={loop}
+                tabIndex={tabIndex}
+                animation={animation}
+                autoplay={autoplay}
+                delay={delay}
+                navigation={navigation}
+                itemRenderer={itemRenderer}
+                items={dataSource?.items ?? []}
+                onClick={onClick}
+            />
+        );
+    };
+    const renderLoading = (): ReactNode => {
+        return <img src={loadingCircleSvg} className="widget-carousel-loading-spinner" alt="" aria-hidden />;
+    };
+    return loading ? renderLoading() : renderCarousel();
```

Comment:
I think this loading state should be inside the "widget-carousel" div element, otherwise users can't have proper control over it. @diego-antonelli do you agree?

If so, either move the div up here or move this loading state into the `CarouselComponent`

Reply:
Why the should not have control on that? Eventually it appears in DOM and they can have access to it via css.

Reply:
Yes, but the structure we normally keep is that we have one outer element with class "widget-..." and then everything is put inside. This functions as a kind of "scope" so that users can always assume the same structure from our widgets.

Reply:
So now it has the structure which we normally keep. Don't know why it's not showing here, but check latest commit

Reply:
Yes I discussed same with @raffiar yesterday and I think he already introduced a wrapper.

Reply:
Yep, it's already there